src/NUnit.Xml.TestLogger/NUnitXmlTestLogger.cs

```diff
@@ -0,0 +1,501 @@
+        }
+
+        private void InitializeImpl(TestLoggerEvents events, string outputPath)
+        {
+            events.TestRunMessage += TestMessageHandler;
+            events.TestResult += TestResultHandler;
+            events.TestRunComplete += TestRunCompleteHandler;
+
+            outputFilePath = Path.GetFullPath(outputPath);
+
+            lock (resultsGuard)
+            {
+                results = new List<TestResultInfo>();
+            }
+
+            localStartTime = DateTime.Now;
+        }
+
+        /// <summary>
+        /// Called when a test message is received.
+        /// </summary>
+        internal void TestMessageHandler(object sender, TestRunMessageEventArgs e)
+        {
+        }
+
+        /// <summary>
+        /// Called when a test result is received.
+        /// </summary>
+        internal void TestResultHandler(object sender, TestResultEventArgs e)
+        {
+            TestResult result = e.Result;
+
+            if (TryParseName(result.TestCase.FullyQualifiedName, out var typeName, out var methodName, out _))
+            {
+                lock (resultsGuard)
+                {
+                    results.Add(new TestResultInfo(
+                        result,
+                        typeName,
+                        methodName));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Called when a test run is completed.
+        /// </summary>
+        internal void TestRunCompleteHandler(object sender, TestRunCompleteEventArgs e)
+        {
+            List<TestResultInfo> resultList;
+            lock (resultsGuard)
+            {
+                resultList = results;
+                results = new List<TestResultInfo>();
+            }
+
+            var doc = new XDocument(CreateAssembliesElement(resultList));
+
+            // Create directory if not exist
+            var loggerFileDirPath = Path.GetDirectoryName(outputFilePath);
+            if (!Directory.Exists(loggerFileDirPath))
+            {
+                Directory.CreateDirectory(loggerFileDirPath);
+            }
+
+            using (var f = File.Create(outputFilePath))
+            {
+                doc.Save(f);
+            }
+
+            var resultsFileMessage = string.Format(CultureInfo.CurrentCulture, "Results File: {0}", outputFilePath);
+            Console.WriteLine(resultsFileMessage);
+        }
+
+        private XElement CreateAssembliesElement(List<TestResultInfo> results)
+        {
+            var testSuites = from result in results
+                group result by result.AssemblyPath
+                into resultsByAssembly
+                orderby resultsByAssembly.Key
+                select CreateTestSuiteElement(resultsByAssembly);
+
+            var element = new XElement("test-run", testSuites);
+
+            element.SetAttributeValue("id", 2);
+
+            var resultString = DetermineResultFromDescendants(element, "test-suite");
```

Comment:
Since we are calculating failed test count, can we use fail test count to get `resultString`?

Reply:
Yes, changed result string be determined from failed > 0 or not